conda/common/url.py

```diff
@@ -125,11 +131,116 @@ def path_to_url(path):
     return path
 
 
+url_attrs = (
+    "scheme",
+    "path",
+    "query",
+    "fragment",
+    "username",
+    "password",
+    "hostname",
+    "port",
+)
+
+
+class Url(namedtuple("Url", url_attrs)):
```

Comment:
Hm since we're starting to include typing can we rework this to use `typing.NamedTuple`? See [this SO](https://stackoverflow.com/a/61844762) for why we need a private class for this to work, if we opt for this solution we should include that SO link as a reference. All references to `url_attrs` can use `Url._fields` instead.

```suggestion
class _Url(NamedTuple):
    scheme: str
    path: str
    query: str
    fragment: str
    username: str
    password: str
    hostname: str
    port: int

class Url(_Url):
```

Reply:
Hi @kenodegard , to be honest, I really wanted to use data classes for this, but figured because we still need to support Python 3.6 it wouldn't be a good idea.

Sounds like we will be dropping support for this soon though, right? Maybe it would be better for me to redo this as a data class instead? (NamedTuple is just a little awkward for this use case, IMHO)

Reply:
True, personally I've just been suuuper hesitant to use dataclasses given [benchmarks](https://twitter.com/anthonypjshaw/status/1498912385268350976?s=20&t=9SMAs71S8zmBJL7J65Hspw) finding it to be significantly slower than normal classes.

Since `_asdict` & `_replace` are predefined in `namedtuple`/`NamedTuple` I'm leaning towards not converting this to a `dataclass`.

Yeah, don't worry about supporting 3.6, it'll be officially dropped in the next release.